bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/handler/MideaACHandler.java

```diff
@@ -0,0 +1,1454 @@
+     * The basic check is if the information from Discovery (or the user update)
+     * is valid. Because V2 devices do not require a cloud provider (or token/key)
+     * The check is for the IP, port and deviceID. This method also resets the dropped
+     * commands, disconnects the socket and stops the connection monitor (if these were
+     * running)
+     */
+    @Override
+    public void initialize() {
+        connectionManager.disconnect();
+        getConnectionManager().cancelConnectionMonitorJob();
+        connectionManager.resetDroppedCommands();
+        connectionManager.updateChannel(DROPPED_COMMANDS, new DecimalType(connectionManager.getDroppedCommands()));
+
+        config = getConfigAs(MideaACConfiguration.class);
+
+        setCloudProvider(CloudProviderDTO.getCloudProvider(config.cloud));
+        setSecurity(new Security(cloudProvider));
+
+        logger.debug("MideaACHandler config for {} is {}", thing.getUID(), config);
+
+        if (!config.isValid()) {
+            logger.warn("Configuration invalid for {}", thing.getUID());
+            if (config.isDiscoveryNeeded()) {
+                logger.warn("Discovery needed, discovering....{}", thing.getUID());
+                updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING,
+                        "Configuration missing, discovery needed. Discovering...");
+                MideaACDiscoveryService discoveryService = new MideaACDiscoveryService();
+
+                try {
+                    discoveryService.discoverThing(config.ipAddress, this);
+                } catch (Exception e) {
+                    logger.error("Discovery failure for {}: {}", thing.getUID(), e.getMessage());
+                }
+                return;
+            } else {
+                logger.debug("MideaACHandler config of {} is invalid. Check configuration", thing.getUID());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Invalid MideaAC config. Check configuration.");
+                return;
+            }
+        } else {
+            logger.debug("Configuration valid for {}", thing.getUID());
+        }
+
+        ipAddress = config.ipAddress;
+        ipPort = config.ipPort;
+        deviceId = config.deviceId;
+        version = Integer.parseInt(config.version);
+
+        logger.debug("IPAddress: {}", ipAddress);
+        logger.debug("IPPort: {}", ipPort);
+        logger.debug("ID: {}", deviceId);
+        logger.debug("Version: {}", version);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        connectionManager.connect();
+    }
+
+    @Override
+    public void discovered(DiscoveryResult discoveryResult) {
+        logger.debug("Discovered {}", thing.getUID());
+        String deviceId = discoveryResult.getProperties().get(CONFIG_DEVICEID).toString();
+        String ipPort = discoveryResult.getProperties().get(CONFIG_IP_PORT).toString();
```

Comment:
This fixes the null warning fot the `get` and can also be applied to the other null warninggs in this file (except the one in line 1448, see other comment)
```
Potential null pointer access: The method get(Object) may return null
```

Also needs the import:
`import java.util.Objects;`

```suggestion
        String deviceId = Objects.requireNonNull(discoveryResult.getProperties().get(CONFIG_DEVICEID).toString());
```

Reply:
I ended up with this prior to your review
```
public void discovered(DiscoveryResult discoveryResult) {
        logger.debug("Discovered {}", thing.getUID());
        Map<String, Object> discoveryProps = discoveryResult.getProperties();
        Configuration configuration = editConfiguration();

        @Nullable
        Object propertyDeviceId = discoveryProps.get(CONFIG_DEVICEID);
        if (propertyDeviceId != null) {
            configuration.put(CONFIG_DEVICEID, propertyDeviceId.toString());
        }

        @Nullable
        Object propertyIpPort = discoveryProps.get(CONFIG_IP_PORT);
        if (propertyIpPort != null) {
            configuration.put(CONFIG_IP_PORT, propertyIpPort.toString());
        }

        updateConfiguration(configuration);

        properties = editProperties();

        @Nullable
        Object propertyVersion = discoveryProps.get(PROPERTY_VERSION);
        if (propertyVersion != null) {
            properties.put(PROPERTY_VERSION, propertyVersion.toString());
        }

        @Nullable
        Object propertySN = discoveryProps.get(PROPERTY_SN);
        if (propertySN != null) {
            properties.put(PROPERTY_SN, propertySN.toString());
        }

        @Nullable
        Object propertySSID = discoveryProps.get(PROPERTY_SSID);
        if (propertySSID != null) {
            properties.put(PROPERTY_SSID, propertySSID.toString());
        }

        @Nullable
        Object propertyType = discoveryProps.get(PROPERTY_TYPE);
        if (propertyType != null) {
            properties.put(PROPERTY_TYPE, propertyType.toString());
        }

        updateProperties(properties);

        initialize();
    }
```

Reply:
Indeed that is another way of doing this, it also solves the warning.
However it does introduce verbosity, both for the annotations and also for the checks.

Reply:
I've created a new comment to continue there. As this has been committed now.